src/librustc_mir/build/expr/as_place.rs

```diff
@@ -6,13 +6,48 @@ use crate::build::{BlockAnd, BlockAndExtension, Builder};
 use crate::hair::*;
 use rustc::mir::interpret::{PanicInfo::BoundsCheck};
 use rustc::mir::*;
-use rustc::ty::{CanonicalUserTypeAnnotation, Variance};
+use rustc::ty::{CanonicalUserTypeAnnotation, Ty, Variance};
 
 use rustc_data_structures::indexed_vec::Idx;
 
+#[derive(Clone)]
+struct PlaceBuilder<'tcx> {
+    base: PlaceBase<'tcx>,
+    projection: Vec<PlaceElem<'tcx>>,
+}
+
+impl PlaceBuilder<'tcx> {
+    fn as_place(self) -> Place<'tcx> {
+        Place {
+            base: self.base,
+            projection: self.projection.into_boxed_slice(),
+        }
+    }
+
+    fn field(&mut self, f: Field, ty: Ty<'tcx>) {
```

Comment:
I think all of these should take `self` and return `Self` in order to make the use-site simpler.

This can be made simpler by also adding a `project` function that pushes an arbitrary `PlaceElem`, so no code has to be repeated in the specialized projection functions